linalg/solvers.cpp

```diff
@@ -766,7 +767,7 @@ void CGSolver::Mult(const Vector &b, Vector &x) const
       return;
    }
    r0 = std::max(nom*rel_tol*rel_tol, abs_tol*abs_tol);
-   if (nom <= r0)
+   if (Monitor(0, nom, r, x) || nom <= r0)
```

Comment:
I think we always should call the monitor with the initial state even when `nom < 0`. To this end, we can keep the `Monitor(...)` call where it is now, and check for monitor convergence here in the `if` statement with an inline method like `bool MonitorConverged() const { return monitor && monitor->HasConverged(); }`.

Also, the when we hit the final iteration (even if that final iteration is 0) we should call the monitor a second time with the same parameters except `final = true`. This allows monitors to properly format output at the end of a solve. I think we need to ensure this behavior not only here but in all iterative solvers, at all the places when they return from `Mult`.

So, in particular, in this context, we should call `Monitor(...)` with `final = false` in both the above case `if (nom < 0.0)` and here in the `if (...)` body.

Finally: the second parameter of `Monitor` is called `norm` and here we pass-in `nom` which is the square of a norm -- we should take a square root of `nom` and if it is negative, we can use either `-sqrt(-nom)` or `-1.0` to indicate an error in the `norm` computation. We should also do something similar to initialize `initial_norm` when `nom < 0.0`.